File: C/http-server-form-file-upload/http.c
#include <unistd.h>
#include <stdio.h>
#include <string.h>

#include "http.h"

void readLineCRLF(int sock, char *line) {
	char *aux=line;
	for(;;) {
		read(sock,aux,1);
		if(*aux=='\n') {
			*aux=0;return;
			}
		else
		if(*aux!='\r') aux++;
		}
	}

void writeLineCRLF(int sock, char *line) {
	char *aux=line;
	while(*aux) {write(sock,aux,1); aux++;}
	write(sock,"\r\n",2);
	}



void sendHttpResponseHeader(int sock, char *status, char *contentType, int contentLength) {
	char aux[200];
	sprintf(aux,"%s %s",HTTP_VERSION,status);
	writeLineCRLF(sock,aux);
	sprintf(aux,"Content-type: %s",contentType);
	writeLineCRLF(sock,aux);
	sprintf(aux,"Content-length: %d",contentLength);
	writeLineCRLF(sock,aux);
	writeLineCRLF(sock,HTTP_CONNECTION_CLOSE);
	writeLineCRLF(sock,"");
	}

int sendHttpResponse(int sock, char *status, char *contentType, char *content, int contentLength) {
	int done, todo;
	char *aux;
	sendHttpResponseHeader(sock, status, contentType, contentLength);
	aux=content; todo=contentLength;
	while(todo) {
		done=write(sock,aux,todo);
		if(done<1) return(0);
		todo=todo-done;
		aux=aux+done;
		}
	return(1);
	}

void sendHttpStringResponse(int sock, char *status, char *contentType, char *content) {
	sendHttpResponse(sock,status,contentType,content,strlen(content));
	}



void sendHttpFileResponse(int sock, char *status, char *filename) {
	FILE *f;
	char *aux;
	char line[200];
	int done;
	long len;
	char *contentType="text/html";

	f=fopen(filename,"r");
	if(!f) {
		sendHttpStringResponse(sock, "404 Not Found", contentType, "<html><body><h1>404 File not found</h1></body></html>");
		return;
		}
	aux=filename+strlen(filename)-1;
	while(*aux!='.' && aux!=filename) aux--;
	if(*aux=='.')
		{
		if(!strcmp(aux,".pdf")) contentType="application/pdf";
		else
		if(!strcmp(aux,".js")) contentType="application/javascript";
		else
		if(!strcmp(aux,".txt")) contentType="text/plain";
		else
		if(!strcmp(aux,".gif")) contentType="image/gif";
		else
		if(!strcmp(aux,".png")) contentType="image/png";
		}
	else
		contentType="application/x-binary";
	
	fseek(f,0,SEEK_END);
	len=ftell(f);
	if(!status) status="200 Ok";
	sendHttpResponseHeader(sock, status, contentType, len);
	rewind(f);
	do {
		done=fread(line,1,200,f);
		if(done>0) write(sock,line,done);
		}
	while(done);
	fclose(f);
	}

File: JAVA/http-server-form-file-upload/HTTP.java
import java.io.*; 
import java.net.*; 

class HTTP {    
	static private final String HTTP_VERSION ="HTTP/1.0";
	static private final String HTTP_CONNECTION_CLOSE ="Connection: close";
	static private final String HTTP_CRLF ="\r\n";
	
	static String readLineCRLF(DataInputStream sIn) {
		byte[] line = new byte[300];
		int p=0;
		try {
  			for(;;) {
				sIn.read(line,p,1);
        			if(line[p]=='\n') return(new String( line, 0, p));
        			else
        				if(line[p]!='\r') p++;
        			}
			}
		catch(IOException ex) { System.out.println("READ IOException"); }
		return(null);
		}

	static void writeLineCRLF(DataOutputStream sOut, String line) {
		try {
			sOut.write(line.getBytes(),0,(byte)line.length());
			sOut.write(HTTP_CRLF.getBytes(),0,2);
			} 
		catch(IOException ex) { System.out.println("WRITE IOException"); }
		}	


	static void sendHttpResponseHeader(DataOutputStream sOut, String status, String contentType, int contentLength) {
		writeLineCRLF(sOut, HTTP_VERSION + " " + status);
		writeLineCRLF(sOut, "Content-Type: " + contentType);
		writeLineCRLF(sOut, "Content-Length: " + contentLength);
		writeLineCRLF(sOut, HTTP_CONNECTION_CLOSE);
		writeLineCRLF(sOut, "");
		}

	static void sendHttpResponse(DataOutputStream sOut, String status, String contentType, 
			byte[] content, int contentLength) {
		sendHttpResponseHeader(sOut,status,contentType,contentLength);
		try {
			sOut.write(content,0,contentLength);
			}
		catch(IOException ex) { System.out.println("IOException"); }
		}


	static void sendHttpStringResponse(DataOutputStream sOut, String status, String contentType, String content) {
		sendHttpResponse(sOut,status,contentType,content.getBytes(),content.length());
		}


	static void sendHttpFileResponse(DataOutputStream sOut, String status, String filePath) {
		String responseStatus="200 Ok";
		String contentType="text/html";
		File f;

		f=new File(filePath);
		if(!f.exists()) {
			sendHttpStringResponse(sOut, "404 Not Found", contentType, 
				"<html><body><h1>404 File not found</h1></body></html>");
			return;
			}
        	if(filePath.endsWith(".pdf")) contentType="application/pdf";
        	else
        	if(filePath.endsWith(".js")) contentType="application/javascript";
        	else
        	if(filePath.endsWith(".txt")) contentType="text/plain";
        	else
        	if(filePath.endsWith(".gif")) contentType="image/gif";
        	else
        	if(filePath.endsWith(".png")) contentType="image/png";

		if(status!=null) responseStatus=status;

		int len= (int) f.length();
		sendHttpResponseHeader(sOut,responseStatus,contentType,len);
		byte[] data = new byte[300];
		int done, readNow;
		InputStream fReader;
		try {
			fReader = new BufferedInputStream(new FileInputStream(f));
			do {
				if(len>300) readNow=300; else readNow=(int)len;
				try {
					done=fReader.read(data,0,readNow);
					len=len-done;
					sOut.write(data,0,done);
					}
				catch(IOException ex) { System.out.println("IOException"); }
				}
			while(len>0);

			}
		catch(FileNotFoundException ex)  { System.out.println("FILE OPEN FileNotFoundException"); }
		}


	}



